Aggiunta capitolo Refactoring (#199)
Il-Libro-Open-Source/book: docs/it/refactoring.md

# Refactor

## Che cos'è il refactoring

Alcuni direbbero che il refactoring è l'arte di crearsi problemi.

> Il refactoring è un insieme di attività dedite a rivedere una o più parti del codice con cui è composta la nostra applicazione.

Reviewer: ```suggestion
Il refactoring è un insieme di attività dedite alla revisione di una o più parti del codice di cui è composta la nostra applicazione.
```

Può interessare un singolo metodo, una singola classe, un intero modulo.

> L'obiettivo di queste attività è rivedere la logica piuttosto che la scrittura del codice che la compone, in modo da ottenere migliori performance (o un maggior disaccoppiamento, una maggiore testabilità) senza però, in alcun modo, alterarne il comportamento e, soprattutto, aggiungere nuove funzionalità.

Reviewer 2: ```suggestion
L'obiettivo di queste attività è rivedere la logica oppure la scrittura del codice che la compone, in modo da ottenere migliori performance (o un maggior disaccoppiamento, una maggiore testabilità) senza però, in alcun modo, alterarne il comportamento e, soprattutto, aggiungere nuove funzionalità.
```

Se l'attività è stata fatta bene, al termine avremo un codice più pulito, comprensibile, di cui non è stato alterato il comportamento e che passa perfettamente tutti i test già presenti.
Ma andiamo per gradi.

## Benefici

> Uno dei principali benefici che otterremo con il refactoring è la rimozione di porzioni di codice non performanti, sostituiti da codice più trasparente e chiaro. E con questo, un'evidente riduzione del debito tecnico accumulato.

Reviewer 3: ```suggestion
Uno dei principali benefici che otterremo con il refactoring è la rimozione di porzioni di codice non performanti, sostituiti da codice più trasparente e chiaro. E con questo, un'evidente riduzione del debito tecnico accumulato.
```

Forse la seconda frase ha un typo o comunque qualcosa da correggere, mi pare che una frase non possa cominciare con la E, a parte questa piccola correzione concettualmente ci siamo!

Reviewer 4: Forse manca il plurale (_questo_ si riferisce effettivamente a _benefici_).
Riguardo l'iniziare una frase con una congiunzione, è perfettamente [corretto](https://accademiadellacrusca.it/it/consulenza/sulluso-delle-congiunzioni-allinizio-di-frase/25).

> Un codice scritto bene non solo aiuta chi lo ha scritto a poterlo rivedere a distanza di qualche tempo senza troppi mal di testa, ma, con il crescere del software e l'introduzione di nuove funzionalità, rende più semplice la vita a chi in futuro dovrà interfacciarsi con esso.

Reviewer: ```suggestion
Un codice scritto bene non solo aiuta chi lo ha scritto a poterlo rivedere a distanza di tempo senza troppi mal di testa ma, con il crescere del software e l'introduzione di nuove funzionalità, rende più semplice la vita a chi in futuro dovrà interfacciarsi con esso.
```

> Un codice scritto bene, non solo aiuta a chi lo ha scritto di poterlo rivedere a distanza di qualche tempo senza troppi mal di testa ma, con il crescere del software e l'introduzione di nuove funzionalità, rende più semplice la vita ai futuri sviluppatori che dovranno interfacciarsi con esso.

Reviewer 2: ```suggestion
Un codice scritto bene non solo aiuta chi lo ha scritto a poterlo rivedere a distanza di qualche tempo senza troppi mal di testa, ma, con il crescere del software e l'introduzione di nuove funzionalità, rende più semplice la vita a chi in futuro dovrà interfacciarsi con esso.
```

> Da un codice scritto bene, si possono ottenere migliori performance. Magari non direttamente ma, una buona attività di refactoring ci permetterà di avere molto più chiaro come il software interagisce, dandoci la possibilità di individuare eventuali colli di bottiglia, anche a livello architetturale.

Reviewer 2: ```suggestion
Da un codice scritto bene si possono ottenere migliori performance. Magari non direttamente, ma una buona attività di refactoring ci permetterà di avere molto più chiaro come il software interagisce, dandoci la possibilità di individuare eventuali colli di bottiglia, anche a livello architetturale.
```

## Quando

> La scrittura del codice non è del tutto lineare, soprattutto se fatta a più mani. Ogni developer ha un suo stile, delle sue convenzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.

Author: ```suggestion
La scrittura del codice non è del tutto lineare, soprattutto se fatta a più mani. Ogni developer ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.
```

> La scrittura del codice non è del tutto lineare, soprattutto se fatta a più mani. Ogni developer ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.

Reviewer 3: ```suggestion
La scrittura del codice non è del tutto lineare, soprattutto se fatta a più mani. Ogni developer ha un suo stile, delle sue convinzioni e un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.
```
Di norma l'ultimo elemento di una serie di virgole dovrebbe avere la e come congiunzione 😄

> La scrittura del codice non è del tutto lineare, soprattutto se scritto a più mani. Ogni developer ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.

Reviewer: ```suggestion
La scrittura del codice non è del tutto lineare, soprattutto se fatta a più mani. Ogni developer ha un suo stile, delle sue convenzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.
```

Reviewer 4: Qui _convinzioni_ era inteso nel senso appropriato del termine; a voler indicare appunto un modo di operare perché si è convinti che sia giusto così, andando contro a quelle che sono le _convenzioni_

Author: Mi torna @mrkrash , ho revertato il cambio di lettera

> La scrittura del codice non è del tutto lineare, soprattutto se scritto a più mani. Ogni sviluppatore ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.

Reviewer 2: ```suggestion
La scrittura del codice non è del tutto lineare, soprattutto se scritto a più mani. Ogni developer ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.
```

Sono diverse le occasioni in cui è possibile fare refactoring e, in alcune di queste, non solo è consigliato ma obbligatorio.

> Poniamo il caso che abbiamo una parte del nostro software con del codice "legacy", magari scritto rispettando dei pattern che, nel tempo, si sono rivelati sbagliati. Oppure scritto di fretta, senza rispettare i principi SOLID e con metodi e variabili che hanno nomi provvisori.

Reviewer: ```suggestion
Poniamo il caso di avere una parte del nostro software con del codice "legacy", magari scritto rispettando dei pattern che, nel tempo, si sono rivelati sbagliati. Oppure scritto di fretta, senza rispettare i principi SOLID e con metodi e variabili che hanno nomi provvisori.
```

O ancora, il codice scritto risulta difficile da testare, perché presenta metodi o classi poco disaccoppiate. Con metodi privati che non fanno attività atomiche.
Fatta una rapida analisi potremmo notare che magari ci sono dei cicli ripetuti, o parti di codice che potrebbero essere evitate se solo avessimo posto dei semafori a monte.

> Con del buon refactoring potremmo rivedere il codice, eliminare delle dipendenze inutili o utilizzare dei costrutti del linguaggio più adatti, con il risultato di migliorarne la lettura, aderire agli standard del linguaggio (che, nell'ottica di rendere il nostro codice più condivisibile possibile, non fa mai male) e, perché no, migliorarne le performance e eliminare debito tecnico.

Reviewer 2: ```suggestion
Con del buon refactoring potremmo rivedere il codice, eliminare delle dipendenze inutili o utilizzare dei costrutti del linguaggio più adatti, con il risultato di migliorarne la lettura, aderire agli standard del linguaggio (che, nell'ottica di rendere il nostro codice più condivisibile possibile, non fa mai male) e, perché no, migliorarne le performance ed eliminare debito tecnico.
```

Come detto, le occasioni non mancano.
Potremmo cogliere l'occasione di una Code Review o l'aggiunta di una nuova feature. Prima di iniziarne lo sviluppo, sicuramente è buona pratica leggere il codice in cui questa nuova funzionalità andrà a integrarsi e, se quest'ultimo risulta di difficile comprensione, è un'ottima occasione per effettuarne il refactoring. Questo ci permetterebbe di ottenere una maggiore conoscenza del codice.

> Un'altra, forse la migliore, dove è (quasi) obbligatorio, è la risoluzione di un bug in una parte di codice convulso, la cui sola lettura ci fa venire il mal di testa. Un'attività di refactoring potrebbe facilmente portare alla luce l'errore presente nel codice.

Reviewer 3: ```suggestion
Un'altra, forse la migliore, dove è (quasi) obbligatorio, è la risoluzione di un bug in una parte di codice convulso, la cui sola lettura ci fa venire il mal di testa. Un'attività di refactoring potrebbe facilmente portare alla luce l'errore presente nel codice.
```

Ho capito cosa intendi dire e secondo me è un buon concetto da portare, la frase però credo sia da rivedere perchè mi sembra poco chiara, forse anche perchè c'è qualche virgola di troppo

Reviewer 4: Forse ce n'è una di troppo (infrange la regola del limite delle 3 virgole per periodo) ma, anche eliminandola, non renderebbe più chiaro il passaggio.
Analizzo nuovamente la frase e vedo se può essere resa più accessibile.

Author: @mrkrash @BrianAtzori  potrebbe andare?

```suggestion
Un'altra occasione potrebbe essere la risoluzione di un bug in una porzione di codice complessa, la cui sola lettura ci fa venire il mal di testa. Un'attività di refactoring potrebbe facilmente portare alla luce l'errore presente nel codice, e spesso risulta la strada migliore, in quanto porta non solo ad una risoluzione, ma anche ad un codice migliore e più mantenibile in futuro.
```

Author: Pensavo di aggiungere anche "riducendo il rischio di regressioni" ma non vorrei sovraccomplicare.

Reviewer 4: mi piace

Reviewer 3: Direi che è ottimo, anche l'aggiunta della riduzioni di rischi secondo me ci può stare!

L'attività di refactoring può essere paragonata a una pulizia profonda di casa, piuttosto che a una routine di sistemazione della propria cameretta, necessaria per mantenere il codice sicuro, di valore ed efficiente.

## Quando è meglio evitare

Non sempre è utile effettuare refactoring.
Ci sono casi in cui il codice è già perfettamente ottimizzato anche se, a una vista inesperta, potrebbe risultare di difficile lettura.

> Ci sono casi in cui le performance sono fondamentali e i principi SOLID non sono sempre rispettati. In questo caso, il refactoring ci porterebbe a creare dei layer aggiuntivi di codice che, sebbene ci renderebbe il codice più comprensibile, avrebbe un impatto negativo sulle performance. Un esempio sono le librerie che permettono le connessioni con delle basi di dati, dove magari è fondamentale la gestione della latenza.

Reviewer: ```suggestion
Ci sono casi in cui le performance sono fondamentali e i principi SOLID non sono sempre rispettati. In questo caso, il refactoring ci porterebbe a creare dei layer aggiuntivi di codice che, sebbene possa rendere il codice più comprensibile, avrebbe un impatto negativo sulle performance. Un esempio sono le librerie che permettono le connessioni con delle basi di dati, dove magari è fondamentale la gestione della latenza.
```

In questi casi, ma più in generale, è necessario tenere a mente che l'attività di refactoring ha tra gli obiettivi quello di rivedere il codice rendendolo più mantenibile, nonché rifattorizzabile.
Diretta conseguenza di questo obiettivo, è la sua applicazione nella pratica del TDD (Test Driven Development), dove il refactoring copre uno degli step.

## Come

Come si può intuire, con le attività di refactoring andremo a migliorare il codice in modo da renderlo meno complesso, più mantenibile, più resiliente.
Per farlo, potremmo iniziare con:

> - Red-Green-Refactor;
> - Rinominare i metodi oscuri;
> - Cambiare la firma dei metodi;
> - Estrarre nuovi metodi;
> - Semplificare i metodi esistenti;
> - Astrarre.

Reviewer: Per sfruttare al meglio md

```suggestion
* Red-Green-Refactor;
* Rinominare i metodi oscuri;
* Cambiare la firma dei metodi;
* Estrarre nuovi metodi;
* Semplificare i metodi esistenti;
* Astrarre.
```

### Red-Green-Refactor

> Forse il metodo più popolare, utilizzato per mettere in pratica il TDD (Test Driven Development), dove lo sviluppatore suddivide il processo di scrittura del codice in tre fasi:

Reviewer 2: ```suggestion
Forse il metodo più popolare, utilizzato per mettere in pratica il TDD (Test Driven Development), dove chi sviluppa suddivide il processo di scrittura del codice in tre fasi:
```

> - Analizza quali sono le esigenze e scrive dapprima il test, che, per ovvi motivi, fallirà (_Red_);
> - Scrive il codice necessario affinché il test passi (_Green_);
> - Lo migliora (_Refactoring_)

Reviewer: ```suggestion
* Analizza quali sono le esigenze e scrive dapprima il test, che, per ovvi motivi, fallirà (_Red_);
* Scrive il codice necessario affinché il test passi (_Green_);
* Lo migliora (_Refactoring_)
```

### Rinominare i metodi oscuri

Ci sono metodi che hanno nomi degni di un codice morse ma che poco dicono sulle operazioni che vanno a intraprendere.

> Un buon inizio potrebbe essere quello di rinominarli in modo che chi legge abbia la possibilità di capire cosa fa quel metodo, senza necessità alcuna di andarlo a verificare.

Reviewer: Che ne pensi se inserissimo degli esempi? Ad esempio uno snippettino con nomi delle variabili tipo "a, b, c" oppure "Pippo" e "Pluto" su un algoritmo banalissimo di facile comprensione a chiunque

Reviewer 4: Sì, assolutamente. Mi adopero per integrarlo

Author: Ciao @mrkrash purtroppo come già ci dicevamo siamo partiti male con sto capitolo perché ho aperto io la PR rendendoti difficile la collaborazione.

Se ti va sentiamoci su LinkedIn per integrare quanto manca!

> Un buon inizio potrebbe essere quello di rinominarli in modo che chi legge, abbia la possibilità di capire cosa fa quel metodo senza necessità alcuna di andarlo a verificare.

Reviewer 2: ```suggestion
Un buon inizio potrebbe essere quello di rinominarli in modo che chi legge abbia la possibilità di capire cosa fa quel metodo, senza necessità alcuna di andarlo a verificare.
```

### Cambiare la firma dei metodi

Partendo dal presupposto che il codice è vivo, si ci può trovare nelle condizioni in cui al metodo inizialmente venivano passate diverse informazioni, che magari ora non sono più necessarie. E magari non lo sono perché nel frattempo, quella parte di codice era stata già snellita in qualche attività precedente.
Possiamo quindi rimuovere le informazioni in eccesso, modificandone la firma.

> Inoltre, nei linguaggi tipizzati, potremmo verificare se c'è la possibilità di rendere detta firma più stringente, ammettendo e/o restituendo tipi meno generici.

Reviewer: Qui sottolineerei che è importante farlo solo nelle major updates, e che sarebbe preferibile utilizzare in prima battuta delle keywords o delle annotazioni speciali come `@deprecated` per comunicare a chi utilizza quei metodi che quei parametri non saranno più utilizzati a posteriori.

Reviewer 4: Sì, ci sta. Apporto dei cambiamenti al periodo

Author: @mrkrash come sopra!

> Inoltre, potremmo verificare, nei linguaggi tipizzati, se c'è la possibilità di rendere detta firma più stringente, ammettendo e/o restituendo tipi meno generici.

Reviewer 2: ```suggestion
Inoltre, nei linguaggi tipizzati, potremmo verificare se c'è la possibilità di rendere detta firma più stringente, ammettendo e/o restituendo tipi meno generici.
```

### Estrarre nuovi metodi

Principalmente si tratta di individuare frammenti di codice che effettuano operazioni contingentate e che possono essere estratte dal codice principale e inserite in una funzione a se stante.
Questo permette di avere un codice meno complesso da interpretare, soprattutto se la nomenclatura dei nuovi metodi è sufficientemente chiara.

### Semplificare i metodi esistenti

Quando il codice infrange la regola _KISS_ (Keep It Simple, Stupid).
Mentre tentiamo l'approccio dell'estrazione, ci rendiamo conto che ci sono parti di codice troppo complesse, piene di condizionalità o che generano ripercussioni su diverse parti del codice.
In questo caso, sarà necessario intervenire dapprima sul rendere meno complesso possibile il codice, rivedendo le condizionalità rendendole più semplici e chiare, magari anche rivedendo come il metodo in oggetto interagisce con il contesto circostante.

### Astrarre

> Spesso, molte delle classi o dei metodi che vengono creati, hanno qualche parte di codice in comune, come le firme dei metodi o l'inizializzazione del contesto.

Reviewer 2: ```suggestion
Spesso, molte delle classi o dei metodi che vengono creati hanno qualche parte di codice in comune, come le firme dei metodi o l'inizializzazione del contesto.
```

> Sebbene questo potrebbe non essere sempre possibile, dopo avere estratto i metodi, ci potremmo trovare dinanzi a molti di questi che sono simili se non uguali. Con l'astrazione possiamo creare delle classi che contengono tali metodi. Metodi che potrebbero divenire ancora più semplici e generici, nonchè facili da interpretare.

Reviewer: ```suggestion
Sebbene questo potrebbe non essere sempre possibile, dopo avere estratto i metodi, ci potremmo trovare dinanzi a funzioni con comportamenti sovrapponibili. Con l'astrazione possiamo creare delle classi che contengono tali metodi. Metodi che potrebbero divenire ancora più semplici e generici, nonché facili da interpretare.
```

## Un Esempio

> Un caso semplice ma spesso presente nel nostro codice, potrebbe essere quello di migliorare le performance rimuovendo le variabili temporanee (il codice che segue è pseudo-codice):

Reviewer 2: ```suggestion
Un caso semplice, ma spesso presente nel nostro codice, potrebbe essere quello di migliorare le performance rimuovendo le variabili temporanee (il codice che segue è pseudo-codice):
```

```
function applyDiscount() {
  basePrice = cart.totalPrice();
  return basePrice * discount();
}
```

Il codice di partenza potrebbe già andare bene così ma, volendolo migliorare, potremmo dichiare la variabile temporanea in modo da facilitare al compilatore la vita, ovvero:

```
function applyDiscount() {
  final basePrice = cart.totalPrice();
  return basePrice * discount();
}
```

> Con questa modifica, non abbiamo cambiato in alcun modo il comportamento. Ovviamente otterremo un piccolo vantaggio sulle performance, a patto che il linguaggio che stiamo usando possa trarre vantaggi dalla differenza tra variabili mutabili e non.

Reviewer 2: ```suggestion
Con questa modifica non abbiamo cambiato in alcun modo il comportamento. Ovviamente otterremo un piccolo vantaggio sulle performance, a patto che il linguaggio che stiamo usando possa trarre vantaggi dalla differenza tra variabili mutabili e non.
```

Ma, riguardando ancora un attimo il codice, potremmo apportare un'altra modifica:

```
function applyDiscount() {
  return cart.totalPrice() * discount();
}
```

Rimuovendo la creazione di una variabile, che oggettivamente risulta solo superflua, andiamo a migliorare sicuramente le performance, non abbiamo introdotto alcuna complessità e la lettura del codice è ancora ampiamente comprensibile.

## Costi

> Lo sviluppo di un software ha un costo, ovviamente anche il refactoring ne ha uno. Ma, fatto con regolarità, sfruttando tutte le occasioni possibili, ovvero effettuando micro-interventi di refactoring, non solo il suo costo è trascurabile ma, complessivamente, abbassa i costi grazie alla maggiore manutenibilità, scalabilità e performance ottenute.

Reviewer: Qui hai usato manutenibilità mentre in precedenza mantenibilità. Personalmente mi piace il primo perché ha un significato più diretto, ma direi comunque di uniformarlo per tutto il documento.

```suggestion
Lo sviluppo di un software ha un costo, ovviamente anche il refactoring ne ha uno. Ma, fatto con regolarità, sfruttando tutte le occasioni possibili, ovvero effettuando micro-interventi di refactoring, non solo il suo costo tende a diventare trascurabile ma, complessivamente, abbassa i costi grazie alla maggiore manutenibilità, scalabilità e performance ottenute.
```

## Refactoring vs Riscrittura

Si potrebbe obiettare che riscrivere una parte del nostro software, basandosi sull'esperienza acquisita, è più veloce o forse più auspicabile del refactoring.
In realtà questa affermazione è fallace: riscrivere significa affrontare un'attività che comporta del tempo non sempre definito, che apporta delle modifiche al funzionamento del software, che andrà testato probabilmente con dei nuovi test e che, non da ultimo, andrà rilasciato in blocco.

> Con il refactoring, dove il dictat del non modificare in alcun modo il comportamento ci impone che i test devono continuare a funzionare senza alcun nostro intervento, è un'attività contingentata che può essere fatta durante lo sviluppo di una nuova funzionalità, se non durante la scrittura dei test stessi. E, soprattutto, il lavoro può essere rilasciato durante il normale ciclo di vita del software, annegato tra i vari rilasci e/o correzioni.

Reviewer: Rivedrei questa frase che non mi suona benissimo

Reviewer 4: Potrebbe sembrare contorta. La rileggo di nuovo e vedo se riesco a riformulare il concetto in maniera più chiara

## Automazione

> Esistono ovviamente diversi strumenti pe automatizzare le procedure di refactoring, spesso integrati direttamente nell'IDE. Forse il più famoso è Sonarqube.

Reviewer 3: ```suggestion
Esistono ovviamente diversi strumenti per automatizzare le procedure di refactoring, spesso integrati direttamente nell'IDE. Forse il più famoso è Sonarqube.
```
Piccola correzione

Questi strumenti, basati su motori di analisi affidabili e complessi, riescono a individuare dalla banale duplicazione del codice alla mancata adozione di buone pratiche, spesso specifiche del linguaggio di programmazione adottato.

> A volte riescono a fare miracoli ma, a noi preme focalizzarci su un punto della questione: l'effort.

Reviewer 2: ```suggestion
A volte riescono a fare miracoli, ma a noi preme focalizzarci su un punto della questione: l'effort.
```

> Quanto effort occorre ad effettuare del refactoring con questi strumenti. O meglio, quanto semplificano la vita di chi sviluppa.

Reviewer: ```suggestion
Quanto effort occorre ad effettuare del refactoring con questi strumenti? O meglio, quanto semplificano la vita di chi sviluppa?
```

> Quanto effort occorre ad effettuare del refactoring con questi strumenti. O meglio, quanto semplificano la vita dello sviluppatore.

Reviewer 2: ```suggestion
Quanto effort occorre ad effettuare del refactoring con questi strumenti. O meglio, quanto semplificano la vita di chi sviluppa.
```

È innegabile che sono dei validi aiuti e che sono formidabili nel farci notare parti di codice da semplificare a cui, magari, non avevamo nemmeno fatto caso.

> Tuttavia, affidare il refactoring all'automazione può creare delle situazioni in cui il codice modificato è così tanto da rendere complicato il controllo da parte di chi sviluppa, tanto che potrebbe avere la tentazione di prendere in blocco tutte le modifiche. Se poi siamo in presenza di una scarsa copertura di test, la frittata è fatta.

Reviewer: ```suggestion
Tuttavia, affidare il refactoring all'automazione può creare delle situazioni in cui il codice modificato è così tanto, da rendere complicato il controllo da parte di chi sviluppa, tanto che potrebbe avere la tentazione di prendere in blocco tutte le modifiche. Se poi siamo in presenza di una scarsa copertura di test, la frittata è fatta.
```

> Ma, affidare il refactoring all'automazione, può creare delle situazioni in cui il codice modificato è così tanto da rendere complicato il controllo da parte dello sviluppatore, che sarebbe tentato di prendere in blocco tutte le modifiche. Se poi siamo in presenza di una scarsa copertura di test, la frittata è fatta.

Reviewer 2: ```suggestion
Tuttavia, affidare il refactoring all'automazione può creare delle situazioni in cui il codice modificato è così tanto da rendere complicato il controllo da parte di chi sviluppa, tanto che potrebbe avere la tentazione di prendere in blocco tutte le modifiche. Se poi siamo in presenza di una scarsa copertura di test, la frittata è fatta.
```

> C'è poi un altro aspetto da considerare: la specificità. Gli strumenti utilizzati sono perfettamente in grado di identificare un eventuale problema, darci indici di complessità ma non riescono a renderci edotti del codice nel suo complesso. Senza conoscere come le varie parti del codice interagiscono nel contesto, difficilmente riusciremo a individuare eventuali errori logici o colli di bottiglia.

Reviewer: ```suggestion
C'è poi un altro aspetto da considerare: la specificità. Gli strumenti utilizzati sono perfettamente in grado di identificare un eventuale problema e a darci indici di complessità. ma non riescono a renderci edotti del codice nel suo complesso. Senza conoscere come le varie parti del codice interagiscono nel contesto, difficilmente riusciremo a individuare eventuali errori logici o colli di bottiglia.
```

> C'è poi un altro aspetto da considerare: la specificità. Gli strumenti utilizzati sono perfettamente in grado di identificare un eventuale problema, darci indici ci complessità ma non riescono a renderci edotti del codice nel suo complesso. Senza conoscere come le varie parti del codice interagiscono nel contesto, difficilmente riusciremo a individuare eventuali errori logici o colli di bottiglia.

Reviewer 2: ```suggestion
C'è poi un altro aspetto da considerare: la specificità. Gli strumenti utilizzati sono perfettamente in grado di identificare un eventuale problema, darci indici di complessità ma non riescono a renderci edotti del codice nel suo complesso. Senza conoscere come le varie parti del codice interagiscono nel contesto, difficilmente riusciremo a individuare eventuali errori logici o colli di bottiglia.
```

> # Table Of Contents
> 
> - Cos'è il refactoring?
> - Quando fare refactoring?
> - Quando NON fare refactoring?
> 
> # Altre idee
> 
> - Differenza (Di intenti oltre che di tempo ed effort) tra refactoring automatico e manuale

Author: ```suggestion
# Refactor
```
